basler_camera.py

```diff
@@ -33,54 +30,89 @@ def __init__(self, config, bus):
                 raise EnvironmentError("no GigE device found")
 
         self.cam.Open()
-        self.cam.ExposureAuto.SetValue('Continuous')
-#        self.cam.Gamma.SetValue(0.4)
-        self.cam.GainAuto.SetValue("Continuous")
-        #self.cam.GainRaw.SetValue(34)
 
-#        a = self.cam.Gain.GetValue()
-#        print(a)
-        self.bus.register('data')
+        self.cam.ExposureAuto.SetValue('Off')
+        self.cam.ExposureTimeAbs.SetValue(170_000)
+        self.cam.GainAuto.SetValue("Off")
+        self.cam.GainRaw.SetValue(34)
+
+
+        self.bus.register('picture')
+        self.bus.register('metadata')
+
+        self.expo_value = -1
+        self.set_exposure()
 
 
-    def start(self):
-        self.input_thread.start()
+    def set_exposure(self):
+        """ Wait till exposure is adapted - timeout is 20 seconds.
+        """
+        self.cam.Width.SetValue(2304)
+        self.cam.OffsetX.SetValue(2304)
+        self.cam.ExposureAuto.SetValue('Once')
 
-    def join(self, timeout=None):
-        self.input_thread.join(timeout=timeout)
+        self.cam.StartGrabbing(pylon.GrabStrategy_LatestImageOnly)
 
-    def run_input(self):
-        folder_name = "set_{}".format(int(time.time()))
-        os.mkdir(os.path.join(self.storage_path, folder_name))
-        
-        settings_path = os.path.join(self.storage_path, folder_name, "settings.pfs")
-        pylon.FeaturePersistence.Save(settings_path, self.cam.GetNodeMap())
-        
+        for k in range(20):
+            time.sleep(1)
+            new_expo = self.cam.ExposureTimeAbs.GetValue()
+            if abs(new_expo - self.expo_value) < 50:
+                break
+            #print(k, new_expo)
+            self.expo_value = new_expo
+
+
+        self.cam.StopGrabbing()
+        self.cam.OffsetX.SetValue(0)
+        self.cam.Width.SetValue(2*2304)
+        print(self.expo_value)
+        #self.cam.ExposureTimeRaw.SetValue(int(expo_value))
+        self.cam.ExposureAuto.SetValue('Off')
+
+
+    def update(self):
+        dt, channel, data = self.listen()
+        # print(dt, channel, data)
+        if channel != "work_dir":
+            return
+        # data = "/home/ovosad/ovosad_data/test_basler_02"
+        save_path = data
+
+        metadata = {"pictures": {}}
+        ev_values = [1, 0.5, 0.25, 0.125, 2, 4, 8]
+        for ii, ev in enumerate(ev_values):
+            try:
+                self.cam.ExposureTimeAbs.SetValue(self.expo_value * ev)
+                short_name = "im_basler_EV{}.tiff".format(ii)
+                long_name = os.path.join(save_path, short_name)
+                self.take_pic(long_name)
```

Comment:
Tady by se měl vracet obrázek, kdy ten první s ev_value 1 by se měl publikovat. Formát obrázku uint8 (aby to načetlo opencv)